### ISSUE TITLE
Working with an unreliable OIDC provider

### ISSUE THREAD

**Issue by granum41** [boylesoftware/tomcat-oidcauth#44]
Hi all, my application is setup to work with my companies OIDC provider, however this particular provider has been having several unplanned outages in recent months which bring down my application given my Tomcat's usage of this plugin. My Tomcat configuration has multiple ways to log in, in addition to using my OIDC provider + tomcat-oidcauth. 

Is there a way I can configure my provider in tomcat-oidcauth such that my application can start, regardless of if a provider is available? 

Sample error:
```
org.apache.catalina.LifecycleException: OpenIDConnectAuthenticator could not load OpenID Connect Provider configuration.
        at org.bsworks.catalina.authenticator.oidc.BaseOpenIDConnectAuthenticator.startInternal(BaseOpenIDConnectAuthenticator.java:815)
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
        at org.apache.catalina.core.StandardPipeline.startInternal(StandardPipeline.java:176)
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
        at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5056)
        at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
        at org.apache.catalina.core.StandardContext.reload(StandardContext.java:3754)
        at org.apache.catalina.manager.ManagerServlet.reload(ManagerServlet.java:1124)
        at org.apache.catalina.manager.HTMLManagerServlet.reload(HTMLManagerServlet.java:641)
        at org.apache.catalina.manager.HTMLManagerServlet.doPost(HTMLManagerServlet.java:217)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:652)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:733)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.catalina.filters.CsrfPreventionFilter.doFilter(CsrfPreventionFilter.java:211)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.catalina.filters.SetCharacterEncodingFilter.doFilter(SetCharacterEncodingFilter.java:109)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:666)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
        at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:690)
        at org.apache.catalina.authenticator.SingleSignOn.invoke(SingleSignOn.java:291)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:373)
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:868)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1589)
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1160)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:820)
Caused by: java.net.SocketTimeoutException: Read timed out
        at java.net.SocketInputStream.socketRead0(Native Method)
        at java.net.SocketInputStream.socketRead(SocketInputStream.java:127)
        at java.net.SocketInputStream.read(SocketInputStream.java:182)
        at java.net.SocketInputStream.read(SocketInputStream.java:152)
        at com.ibm.jsse2.b.a(b.java:123)
        at com.ibm.jsse2.b.a(b.java:269)
        at com.ibm.jsse2.av.a(av.java:12)
        at com.ibm.jsse2.av.a(av.java:513)
        at com.ibm.jsse2.f.read(f.java:9)
        at java.io.BufferedInputStream.fill(BufferedInputStream.java:257)
        at java.io.BufferedInputStream.read1(BufferedInputStream.java:297)
        at java.io.BufferedInputStream.read(BufferedInputStream.java:356)
        at sun.net.www.http.HttpClient.parseHTTPHeader(HttpClient.java:766)
        at sun.net.www.http.HttpClient.parseHTTP(HttpClient.java:709)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:1605)
        at sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1510)
        at com.ibm.net.ssl.www2.protocol.https.b.getInputStream(b.java:19)
        at org.bsworks.catalina.authenticator.oidc.ConfigProvider.loadDocument(ConfigProvider.java:138)
        at org.bsworks.catalina.authenticator.oidc.ConfigProvider.get(ConfigProvider.java:106)
        at org.bsworks.catalina.authenticator.oidc.OPConfigurationsProvider.getOPConfiguration(OPConfigurationsProvider.java:72)
        at org.bsworks.catalina.authenticator.oidc.BaseOpenIDConnectAuthenticator.startInternal(BaseOpenIDConnectAuthenticator.java:813)
        ... 40 more

```

Much thanks,
Avery

**Comment by levahim** [comment]
Interesting point. We'll consider adding a flag to the authenticator configuration that allows it to start without a given provider (it'll be failing to authenticate using the provider if attempted).

**Comment by levahim** [comment]
This has been addressed in release 2.5.0. See "optional" OP configuration option.
